python/mujinwebstackclient/controllerwebclientraw.py

```diff
@@ -280,3 +363,132 @@ def CallGraphAPI(self, query, variables=None, headers=None, timeout=5.0):
             raise ControllerGraphClientException(_('Unexpected server response %d: %s') % (statusCode, raw), statusCode=statusCode, response=response)
 
         return content['data']
+
+    async def _OpenWebSocketConnection(self):
+        authorization = self._session.auth.GetAuthorizationHeader()
+
+        # URL to http GraphQL endpoint on Mujin controller
+        graphEndpoint = '%s/api/v2/graphql' % self._baseurl
+        parsedUrl = urlparse(graphEndpoint)
+        # parse url and handle different scheme
+        webSocketScheme = ''
+        if parsedUrl.scheme == 'https':
+            webSocketScheme = 'wss'
+        elif parsedUrl.scheme == 'http':
+            webSocketScheme = 'ws'
+        uri = '%s://%s%s' % (webSocketScheme, parsedUrl.netloc, parsedUrl.path)
+
+        # prepare the headers
+        headers = {
+            'Content-Type': 'application/json',
+            'Accept': 'application/json',
+            'X-CSRFToken': 'token',
+        }
+        subprotocols = ['graphql-ws']
+
+        # decide on using unix socket or not
+        adapter = self._session.adapters.get('http://')
+        if isinstance(adapter, UnixSocketAdapter):
+            self._websocket = await websockets.unix_connect(
+                path=adapter.get_unix_endpoint(),
+                uri=uri,
+                subprotocols=subprotocols,
+                additional_headers=headers,
+            )
+        else:
+            self._websocket = await websockets.connect(
+                uri=uri,
+                subprotocols=subprotocols,
+                additional_headers=headers,
+            )
+
+        await self._websocket.send(json.dumps({
+            'type': 'connection_init',
+            'payload': {
+                'Authorization': authorization,
+            }
+        }))
+    
+    async def _ListenToWebSocket(self):
+        try:
+            async for response in self._websocket:
+                data = json.loads(response)
+                if data['type'] == 'connection_ack':
+                    log.debug('received connection_ack')
+                elif data['type'] == 'ka':
+                    # received keep-alive "ka" message
+                    pass
+                else:
+                    # parse to get the subscriptionId so that we can call the correct callback function
+                    subscriptionId = data.get('id')
+                    if subscriptionId in self._subscriptions:
+                        subscription = self._subscriptions[subscriptionId]
+                        subscription.GetSubscriptionCallbackFunction()(data.get('payload') or {})
```

Comment:
@kamieater1240 We discussed about parsing the errors in the GraphQL query result similar to `CallGraphAPI` function but I don't see this logic implemented here

Reply:
Another thing we discussed was sending the data and the errors via separate arguments to the callback function